src/library/dao/macrodao.cpp

```diff
@@ -0,0 +1,111 @@
+#include "library/dao/macrodao.h"
+
+#include <QDebug>
+#include <QSqlQuery>
+
+#include "library/queryutil.h"
+
+bool MacroDAO::saveMacro(TrackId trackId, Macro* macro, int slot) const {
+    QSqlQuery query(m_database);
+    if (macro->getId() == -1) {
```

Comment:
```suggestion
    if (macro->getId() == kNoMacroId) {
```

Reply:
We need a global constant for that, I have seen this in many places. Will file a separate PR after GSoC.

Reply:
`DbId`

Reply:
`kInvalidValue` is private :/

Reply:
fixed in d04d8d8453